[WIP] Graph time format browser-localized with user preference override

davkal opened the pull request:
When viewing longer range graphs, the dates on the x-axis are formatted in the US style: mm/dd. This PR allows more flexibility:

- use browser language to determine formatting
- allow user to override format in user preferences

Fixes #1459

daniellee committed:
#1459 - spike for graph day/month localization

Need to add a config option for day month. 'browser' tries
to use the browser's locale settings.

davkal committed:
Graph timeformat as user preference

- adds `monthDayFormat` to user preferences with dropdown
- stored as `MonthDayFormat` in the `preferences` table

torkelo commented on public/app/plugins/panel/graph/graph.ts:
```diff
@@ -709,6 +710,31 @@ class GraphElement {
+      .replace('20', '%d');
+  }
+
+  toLocaleDateStringSupportsLocales() {
```
could this be renamed to isLocaleDateStringLocalesSupported ,  reading this code I struggled to understand that this function was bool check and not a conversion function

torkelo replied:
agree, canUse is better :+1:

davkal committed:
Review feedback

- renamed function that checks for locales support
- fixed DB column definition to allow NULL as that is the value for UI
  defaults
- added tooltip to option
- renamed label option to keep to single line

davkal commented:
Looks like a TS issue, strange that the linter did not detect it. Working on fix...

davkal committed:
Fix user object in test

torkelo reviewed (changes requested):
Hi, 

This looks like a good start but seems to be missing a few things. I could not get it to work. The contextSrv.user.monthDayFormat is always browser no matter what I set it to in user preferences. Does not look like this property is set for the current user here:

https://github.com/grafana/grafana/blob/master/pkg/api/index.go#L70 

There is one thing that can complicate this and that is the PNG server side rendering, if you use "browser" locale then that could change depending on what locale Grafana server is running in. We currently send in timezone to the server side rendering but not locale. 

Also there is some code there that is in the graph.ts file that should be moved to common util / time lib so other panels like heatmap can also take advantage of this setting.

davkal committed:
Pass time format to frontend, more time formats

davkal commented:
> I could not get it to work. The contextSrv.user.monthDayFormat is always browser no matter what I set it to in user preferences. 

Yeah, the final piece of plumbing was missing, pushed a fix. Also: [LOL](https://github.com/grafana/grafana/blob/25f313c7ececaf0b29b7f6b9cd52cd4159cc906c/public/app/core/services/context_srv.ts#L22)

I'll see if the code can be reused and move some things around.

davkal committed:
Move all time format to utils/ticks.ts

davkal committed:
Pass monthDayFormat via url context service

davkal commented:
@torkelo not sure if this is most elegant way to pass the format parameter around. PTAL

torkelo commented:
This seems to be working, the only problem I feel is that it's a specific option for just day & month. The profile option is named Graph Date Format, it does not provide an option to change year formatting. 

So the only open question is if there is a way to extend this option to also include year formatting. So we do not need another option for that. 

![image](https://user-images.githubusercontent.com/10999/46803019-5e26fa00-cd5f-11e8-9b6e-4eccb94d352b.png)

davkal commented on public/app/core/utils/ticks.ts:
```diff
@@ -208,3 +211,34 @@ export function getStringPrecision(num: string): number {
+
+  const language = navigator.languages ? navigator.languages[0] : navigator.language || 'en-US';
+  return new Date(Date.UTC(1977, 11, 20, 0, 0, 0))
+    .toLocaleDateString(language, { month: 'numeric', day: 'numeric' })
```
Do you know of a concrete recommendation to use one over the other? Otherwise I see no basis to change this.

torkelo replied:
No, i was just this thinking that it might be simpler just read the format directly instead of formating a date and parsing it to reverse engineer the format

andig commented:
> An extended solution would need to specify all formats for ticks, currently we have:

That would be awesome @davkal. 

> Would you have a form field for each or comma separated? Or set up a new date formatting page in user preferences?

From troubled user's perspective I really wouldn't care. A simple form could just combine all settings as @torkelo suggested into a dropdown with reasonable defaults for US plus selected countries. US might look like a concatenated version of:

    "hh:mm:ss.SSS",
    "PT1S: HH:mm:ss",
    "PT1M: HH:mm",
    "PT1H: YYYY-MM-DD",
    "P1DT: YYYY-MM-DD",
    "P1YT: YYYY"

This could end in the config json as suggested by @torkelo. Anything that does not start with a `P` would be the default format. A "custom" entry in the dropdown would allow specifying my own format string, no matter if I need to build that outside grafana UI as its really a one-time task and PRs for additional defaults would easily be done from there on.

Does that input help any?

For reference, here is what the underlying flot library does for time formatting: https://github.com/flot/flot/blob/889f7c97e3580b26e86c927f4507c150006cbcf1/source/jquery.flot.time.js#L386. It includes ability to specify custom names for weekdays and months (nice to have but secondary concern).

andig commented:
@davkal would you be interested in fixing the conflicts and implementing the requested changes? Otherwise I might attempt a fork based on what you've done sofar.

davkal commented:
Not going forward with this, will retry with more comprehensive approach.
